docs/proposals/proposal-ratelimiter.md

```diff
@@ -0,0 +1,162 @@
+
+I think we need to add a new attribute, `PreviousNamespaces`, to
+`BundleInstanceStatus`. It will be an slice of strings. The list will be ordered
+as first in first out. Newer jobs will append to the list, we simply pop the
+first ones off.
+
+```
+  // BundleInstanceStatus status is a service instance status.
+    Bindings           []LocalObjectReference `json:"bindings"`
+    State              State                  `json:"state"`
+    LastDescription    string                 `json:"lastDescription,omitempty"`
+    Jobs               map[string]Job         `json:"jobs"`
+    PreviousNamespaces []string               `json:"previousNamespaces"`
+  }
+```
+
+When we start the action, we should append the namespace to the
+`PreviousNamespaces` field. If `KeepNamespaceOnError` is false, we will delete
+the sandbox namespace and should remove it from the list. This would probably be
+done in a defer similar to how we do the `DestroySandbox` call.
+
+We will keep at most 1 old failed namespace instead of all of them. The idea
+here is to minimize wasting resources in the cluster and still allow developers
+& admins to have access to the failed action. During a deprovision & unbind,
+once we get the `ServiceInstance` we will see if there are any
+`PreviousNamespaces`. If there are old ones, we will delete the first one in the
+list, id 0. Before creating a new sandbox namespace.
+
+TBD: not sure if this should be done at the handler level or further down. I'll
+leave that as an implementation detail.
+
+NOTE: we could make the number of namespaces kept to be configurable if users
+want to keep _n_ number of namespaces. But I think that might add unneeded
+complexity.
+
+### Rate limit the API
+
+The broker will respond to an API call as many times as you call it. Often this
+means launching a new APB which can result in a ton of cluster resources being
+used when the API call was a result of a previous failure. For example, if a
+provision call fails, the platform will likely call a deprovision to prevent any
+dangling resources. If the deprovision fails, the platform will continue to call
+deprovision until it works. If the `KeepNamespaceOnError` is `true`, we will
+keep the namespaces for *every* failed call around wasting precious cluster
+resources.
+
+The proposal above will mitigate the left over namespaces. But if we now a call
+will always fail because it has failed the last n times. There really is no
+point to continue to spawn new APBs and namespaces. We should backoff the calls
+exponentially. We take n calls with no limit, once those n are exhausted we take
+every 3rd call, then we back that off to every 5th call. After m failed attempts
+we stop executing APBs and simply return failure always.
+
+At first I thought we'd have a time limit as well but the exponential backoff
+seems like a better idea to apply across the board.  A new configuration item
+includes a list of action handlers we want to limit.
+
+```yaml
```

Comment:
Do we need a configuration? Or could we simply have an exponential backoff for subsequent requests on something that previously failed.

Thinking about it, what if in the first go we simply rate limited deprovision. No configuration no special cases. If deprovision fails over and over we simply attempt deprovision less and less often.

Reply:
> There is one potential gotcha with the proposal to only keep the most recently-failed namespace for a given operation. We keep the namespace so an admin can inspect it and figure out what went wrong. That person's job will be more difficult if namespaces keep getting deleted while the admin is inspecting them.

I didn't think of the scenario of the admin looking at the specific namespace that may go away from them. I was kind of under the impression that it is probably failing for the same reason constantly, so looking at the most recent one would suffice. But after reading your comment, I can see how they might be looking at the last failed error and POOF, it's gone. Now they have to restart with the new one to get back to where they were then POOF, it's also gone.

> 
> In theory the admin could add a finalizer to a namespace to ensure it doesn't get deleted mid-troubleshooting, but that doesn't feel like the right user experience.

I wouldn't expect them to have to go around the broker to solve this.

> 
> Would it make sense to only keep the oldest one, and delete any newer namespaces that fail in the same way? Or maybe only keep the oldest and newest, but none in between? Or something else?

This is certainly an interesting approach. I'm seeing 2 new attributes: `FirstFailedSandbox` and `LastSandbox`. The former will contain the value of the sandbox that failed or empty if no failure has occurred. The latter will contain the last sandbox run. My first iteration had a list of sandboxes with the oldest being the first one in the list, not sure if we need to keep that feature or not.

@shawn-hurley thinks just keeping the first failed sandbox would suffice (see https://github.com/openshift/ansible-service-broker/pull/1130#issuecomment-439970003) but I think we'll need at least the first failed one and the latest.